### PR TITLE
wxGUI/preferences: allow the user set maximum memory to be used by modules

### LINKED ISSUE #3267: Global memory use setting in preferences
Many GRASS modules have an option to set the maximum memory to be used in processing. This defaults to 300Mb in all cases. Most computers today come with at least 8Gb of RAM, and many are equipped with much more. Usually, the more RAM a module is allowed to use, the faster it runs. 

It would be convenient if the default minimum could be set globally more easily in the wxGUI preferences environment to match the capacity of a user's workstation rather than having to set it individually in every module. While this is doable via g.gisenv, it is pretty opaque to most users on how to do this.

### PR DESCRIPTION
**Pull request by tmszi**
Closes #3267.

1. ~Solution require Python **psutil** lib to be installed.~
2. ~The global memory size in MB used by modules is calculated using the formula `total_memory - (1 / 6 * total_memory)`~

https://github.com/OSGeo/grass/assets/50632337/613819cd-a6bc-4565-9595-39615a90c47f

### CODE REVIEW

**Review by petrasovaa — CHANGES_REQUESTED** [review]
Let's not use psutil, it's another external dependency. Since MEMORYMB is by default not set, I would leave it unset, probably need to use TextCtrl instead of SpinCtrl.

The same could be done for NPROCS variable.

What I am not sure is whether these variables are actually respected by the tools.

**Comment by tmszi** [comment]
> Let's not use psutil, it's another external dependency. 

There is no other cross-platform way to find out the total RAM size of a machine.

> Since MEMORYMB is by default not set, I would leave it unset, probably need to use TextCtrl instead of SpinCtrl.

Could you explain why to use the TextCtrl widget, please?

**Comment by tmszi** [comment]
> Since MEMORYMB is by default not set, I would leave it unset

Without setting gis env `MEMORYMB` is not possible simple dynamically change memory value for all modules.

https://github.com/OSGeo/grass/blob/2841c5f36aafe69c72301cc7f89f95b546caf3b5/lib/gis/parser_standard_options.c#L252-L267

**Comment by wenzeslaus** [comment]
These MEMORYMB and NPROCS are [gisenv variables](https://grass.osgeo.org/grass83/manuals/variables.html#list-of-selected-grass-gisenv-variables), so it seems to me that a analysis is needed to determine how the GUI setting should interact with switching mapsets and turning GRASS GIS on and off. Notably, the settings is already managed by the MEMORYMB variable, so why to involve GUI setting file?

Maybe, this is supposed to be completely different from the GUI settings or at least separate from implementation perspective but perhaps as a tab in the setting dialog (just thinking out-loud here). Maybe, g.gisenv just needs a wrapper.

The default for in the individual tools when MEMORYDB is not set is already there - it is 300.

**Comment by petrasovaa** [comment]
> > Let's not use psutil, it's another external dependency.
> 
> There is no other cross-platform way to find out the total RAM size of a user machine.
> 
> #3267 _It would be convenient if the default minimum could be set globally more easily in the wxGUI preferences environment **to match the capacity of a user's workstation** rather than having to set it individually in every module._

This mixes two things together, setting the default globally and setting it to match the capacity. I am not in favor of the latter, e.g. I have rarely the full capacity (or 5/6 of it) available, and it's not worth adding a new dependency.

> 
> > Since MEMORYMB is by default not set, I would leave it unset, probably need to use TextCtrl instead of SpinCtrl.
> 
> Could you explain why to use the TextCtrl widget, please?

Sorry, I was too brief, my suggestion was to use TextCtrl, which allows you (unlike SpinCtrl) to have no value set. So by default MEMORYMB is not set, so initially the TextCtrl would be empty. SpinCtrl is not useful for input of large numbers anyway I think. We could include integer validator in the TextCtrl I suppose.

**Comment by petrasovaa** [comment]
> These MEMORYMB and NPROCS are [gisenv variables](https://grass.osgeo.org/grass83/manuals/variables.html#list-of-selected-grass-gisenv-variables), so it seems to me that a analysis is needed to determine how the GUI setting should interact with switching mapsets and turning GRASS GIS on and off. Notably, the settings is already managed by the MEMORYMB variable, so why to involve GUI setting file?
> 

We could remove this from the settings file and just keep it in the settings GUI, it would be dynamically read from the MEMORYMB variable when you open settings, but it won't be recorded in the json file.

When switching mapsets, the variable seem to persist, so this looks ok to me.

> Maybe, this is supposed to be completely different from the GUI settings or at least separate from implementation perspective but perhaps as a tab in the setting dialog (just thinking out-loud here). Maybe, g.gisenv just needs a wrapper.
> 
> The default for in the individual tools when MEMORYDB is not set is already there - it is 300.

File: gui/wxpython/gui_core/preferences.py [hunk 1]
```diff
@@ -1245,6 +1272,54 @@ def _createCmdPage(self, notebook):
 
         gridSizer.Add(verbosity, pos=(row, 1), flag=wx.ALIGN_RIGHT)
 
+        row += 1
+        # nprocs
+        gridSizer.Add(
+            StaticText(
+                parent=panel,
+                id=wx.ID_ANY,
+                label=_("Number of threads for parallel computing:"),
```
Comment (petrasovaa) [inline gui/wxpython/gui_core/preferences.py:1281]: I wonder if we should add "...computing (supported tools only): ". Or "Number of cores to be used by parallelized tools".
Similarly, the memory handling is implemented only in certain tools, so maybe "Maximum memory in MB to be used (supported tools only)". But maybe you can come up with a better wording... I just don't want this to be misunderstood.

Comment (tmszi) [reply gui/wxpython/gui_core/preferences.py]: Fixed with https://github.com/OSGeo/grass/pull/3307/commits/97735efa011f9eed5b9105e1cc14ff4a2a3192c6.

**Review by petrasovaa — APPROVED** [review]
This looks good to me!

**Pull request merged**